contracts/base/snx-base/SNXRewardUniLPStrategyWithBuyback.sol

```diff
@@ -0,0 +1,325 @@
+*   3. SHRIMP https://etherscan.io/address/0x9f83883FD3cadB7d2A83a1De51F9Bf483438122e#code
+*   4. BASED https://etherscan.io/address/0x5BB622ba7b2F09BF23F1a9b509cd210A818c53d7#code
+*   5. YFII https://etherscan.io/address/0xb81D3cB2708530ea990a287142b82D058725C092#code
+*   6. YFIGovernance https://etherscan.io/address/0xBa37B002AbaFDd8E89a1995dA52740bbC013D992#code
+*
+*
+*
+*   Respecting the current system design of choosing the best strategy under the vault, and also rewarding/funding
+*   the public key that invokes the switch of strategies, this smart contract should be deployed twice and linked
+*   to the same vault. When the governance want to rotate the crop, they would set the reward source on the strategy
+*   that is not active, then set that apy higher and this one lower.
+*
+*   Consequently, in the smart contract we restrict that we can only set a new reward source when it is not active.
+*
+*/
+
+contract SNXRewardUniLPStrategyWithBuyback is StrategyBase {
+
+  using SafeMath for uint256;
+  using SafeERC20 for IERC20;
+
+  address public uniLPComponentToken0;
+  address public uniLPComponentToken1;
+
+  address public distributionPool;
+  uint256 public buybackRatio;
+
+  bool public pausedInvesting = false; // When this flag is true, the strategy will not be able to invest. But users should be able to withdraw.
+
+  SNXRewardInterface public rewardPool;
+
+  // a flag for disabling selling for simplified emergency exit
+  bool public sell = true;
+  uint256 public sellFloor = 1e6;
+
+  mapping (address => address[]) public uniswapRoutes;
+
+  event ProfitsNotCollected();
+
+  // This is only used in `investAllUnderlying()`
+  // The user can still freely withdraw from the strategy
+  modifier onlyNotPausedInvesting() {
+    require(!pausedInvesting, "Action blocked as the strategy is in emergency state");
+    _;
+  }
+
+  constructor(
+    address _storage,
+    address _underlying,
+    address _vault,
+    address _rewardPool,
+    address _rewardToken,
+    address _uniswapRouterV2,
+    address _distributionPool,
+    uint256 _buybackRatio
+  )
+  StrategyBase(_storage, _underlying, _vault, _rewardToken, _uniswapRouterV2)
+  public {
+    require(IRewardPool(_distributionPool).lpToken() == _vault, "Incompatible pool");
+    require(_buybackRatio <= 10000, "Buyback ratio too high");
+    buybackRatio = _buybackRatio;
+    distributionPool = _distributionPool;
+    uniLPComponentToken0 = IUniswapV2Pair(underlying).token0();
+    uniLPComponentToken1 = IUniswapV2Pair(underlying).token1();
+    rewardPool = SNXRewardInterface(_rewardPool);
+  }
+
+  function depositArbCheck() public view returns(bool) {
+    return true;
+  }
+
+  /*
+  *   In case there are some issues discovered about the pool or underlying asset
+  *   Governance can exit the pool properly
+  *   The function is only used for emergency to exit the pool
+  */
+  function emergencyExit() public onlyGovernance {
+    rewardPool.exit();
```

Comment:
We should only exit if pool balance is greater than 0, since the underlying reward pool would cause a revert in case of a 0.

Reply:
Added